File: ios/MochiShared/Input/AccentMap.swift
import Foundation

/// Alternate characters shown when a key is held.
///
/// Matches the US-English system keyboard's set, in the system's order — users reach these by
/// muscle memory (hold `e`, slide one right, get `é`) and reordering them is a subtle way to make
/// a keyboard feel wrong without anyone being able to say why.
enum AccentMap {
    static func alternates(for character: String) -> [String] {
        table[character.lowercased()] ?? []
    }

    static func hasAlternates(for character: String) -> Bool {
        !alternates(for: character).isEmpty
    }

    private static let table: [String: [String]] = [
        "a": ["à", "á", "â", "ä", "æ", "ã", "å", "ā"],
        "c": ["ç", "ć", "č"],
        "e": ["è", "é", "ê", "ë", "ē", "ė", "ę"],
        "i": ["î", "ï", "í", "ī", "į", "ì"],
        "l": ["ł"],
        "n": ["ñ", "ń"],
        "o": ["ô", "ö", "ò", "ó", "œ", "ø", "ō", "õ"],
        "s": ["ß", "ś", "š"],
        "u": ["û", "ü", "ù", "ú", "ū"],
        "y": ["ÿ"],
        "z": ["ž", "ź", "ż"],

        // Punctuation alternates from the numeric and symbol planes. Included because a user who
        // has learned to hold `?` for `¿` on the system keyboard will try it here.
        "-": ["–", "—", "•"],
        "/": ["\\"],
        "$": ["₹", "€", "£", "¥", "₩"],
        "&": ["§"],
        "\"": ["“", "”", "„", "«", "»"],
        "'": ["’", "‘", "`"],
        ".": ["…"],
        "?": ["¿"],
        "!": ["¡"],
        "%": ["‰"],
        "=": ["≠", "≈"]
    ]
}

File: android/app/src/main/java/com/mochi/keyboard/ime/render/AccentMap.kt
package com.mochi.keyboard.ime.render

/** Long-press accent alternates, in the system keyboard's own order — ported from iOS's
 * `AccentMap.swift` so users reach these by the same muscle memory. */
object AccentMap {
    private val map: Map<String, List<String>> = mapOf(
        "a" to listOf("à", "á", "â", "ä", "æ", "ã", "å", "ā"),
        "e" to listOf("è", "é", "ê", "ë", "ē", "ė", "ę"),
        "i" to listOf("ì", "í", "î", "ï", "ī", "į"),
        "o" to listOf("ò", "ó", "ô", "ö", "õ", "ø", "ō"),
        "u" to listOf("ù", "ú", "û", "ü", "ū"),
        "y" to listOf("ý", "ÿ"),
        "n" to listOf("ñ", "ń"),
        "c" to listOf("ç", "ć", "č"),
        "s" to listOf("ś", "š", "ß"),
        "z" to listOf("ž", "ź", "ż"),
        "l" to listOf("ł"),
        "g" to listOf("ğ"),
        "0" to listOf("°"),
        "1" to listOf("¹"),
        "2" to listOf("²"),
        "3" to listOf("³"),
        "-" to listOf("–", "—"),
        "'" to listOf("'", "'", "`"),
        "\"" to listOf("“", "”"),
        "?" to listOf("¿"),
        "!" to listOf("¡"),
        "$" to listOf("€", "£", "¥", "¢")
    )

    fun alternates(character: String): List<String> = map[character.lowercase()] ?: emptyList()
}
